sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -6681,8 +6777,10 @@ pi_result piEnqueueEventsWaitWithBarrier(pi_queue Queue,
   std::vector<pi_command_list_ptr_t> CmdLists;
 
   // There must be at least one L0 queue.
-  PI_ASSERT(!Queue->ComputeQueueGroup.ZeQueues.empty() ||
-                !Queue->CopyQueueGroup.ZeQueues.empty(),
+  auto InitialComputeGroup = Queue->ComputeQueueGroupsByTID.begin()->second;
+  auto InitialCopyGroup = Queue->CopyQueueGroupsByTID.begin()->second;
+  PI_ASSERT(!InitialComputeGroup.ZeQueues.empty() ||
+                !InitialCopyGroup.ZeQueues.empty(),
```

Comment:
I think you have to count all queues from all threads (the first one may not necessarily have been used)

Reply:
Perhaps you can just move assert after the if/else below where the `CmdLists` computed should be non-empty

Reply:
The ASSERT is checking for a valid/usable queue. If there are no ZeQueues of either compute or copy type, the queue is useless.
The presence of commandlists associated with the L0 queues is based on activity on the queue. That is a different check. Here, we do check all groups of all threads. We tolerate no activity on the queue and create a commandlist if the queue had been unused so far.

Reply:
I mean that a queue may be created in one main thread, and then used from other worker threads. In this case the .begin() may not have ant ZeQueues in it. Actually I am not sure if .begin() is even going to point to the main thread in this case.

Reply:
We are checking if the queue has any associated L0 compute/copy queues. The initial thread may find some and then all subsequent threads will use the same L0 queues. It doesn't matter which entry the "begin" returns. All entries have the same L0 queues. (This is different from commandlists).

Reply:
Hmm, I am now thinking that with immediate command-lists these `ZeQueues` would always be empty

Reply:
When a queue is constructed there are nullptr entries in the computeGroup, totaling the number of CCS's visible. For the copy group, the number of copy engines.

Reply:
OK, it will be non-empty vector of null values, right. Not sure what this assert is checking though then
OK. Let it be

Reply:
Yes, the assert is not very useful. It will only be hit if somehow the device presents no compute or copy queues at all, and the first thing done on such a queue is appendBarrier.

And yes, with immediate commandlists being used for the queue, the ZeQueues will contain only nullptrs. Their only purpose when using immediate commandlists is their size, used to allocate that many immediate commandlists.